### ISSUE TITLE
Error: Invalid operand type and invalid combination of opcode and operands in PortableNetworkGraphic.pas with FPC

### ISSUE THREAD

**Issue by regs01** [graphics32/graphics32#219]
Trying to compile Lazarus package throws error `GR32_PortableNetworkGraphic.pas(4067,26) Error: Invalid operand type`.  Lazarus and FPC are a trunk. Last time I was in Assembler was 20 years ago, so had to Google. So, I suppose that it expect value to be given in square brackets, as LEA accepts an address? Couple of lines above x64 version is given in square brackets. With brackets it will pass, but then there is another error, which I can't understand. 
`GR32_PortableNetworkGraphic.pas(5647,9) Error: Asm: [movzx reg32,mem32] invalid combination of opcode and operands`
on line
`MOVZX   EAX, [EDX]`

**Comment by andersmelander** [comment]
> I suppose that it expect value to be given in square brackets, as LEA accepts an address?

Yes:
```
LEA     EAX, [EAX + ECX]
LEA     EDX, [EDX + ECX]
```
is correct.

and
```
MOVZX   EAX, [EDX]
```
should be
```
MOV     EAX, [EDX]
```

**Comment by regs01** [comment]
Next is even more complicated
`GR32_PortableNetworkGraphic.pas(5643,9) Error: Internal error 2003010802`

on a line 
```asm
{$IFNDEF FPC}
        MOV     EDI, [GCrcTable]
{$ELSE}
        MOV     EDI, [EIP + GCrcTable]
{$ENDIF}  
```

` MOV     EDI, [GCrcTable]` would compile, though don't know what would it do.


Then couple of MOVZX in GR32_Png. All throwing error. Shouldn't be MOVZX in x86 either?
```asm
  MOVZX   EBX, [EAX + 4 * ECX]
  MOVZX   EBX, [EDI + EBX]
  MOV     [EDX + 4 * ECX + $02], BL

  MOVZX   EBX, [EAX + 4 * ECX + $01]
  MOVZX   EBX, [EDI + EBX]
  MOV     [EDX + 4 * ECX + $01], BL

  MOVZX   EBX,[EAX + 4 * ECX + $02]
  MOVZX   EBX,[EDI + EBX]
  MOV     [EDX + 4 * ECX], BL

  MOVZX   EBX, [EAX + 4 * ECX + $03]
  MOV     [EDX + 4 * ECX + $03], BL

  ADD     ECX, 1
  JS      @Start

  POP     EDI
  POP     EBX 
```


x64 compiles without errors

**Comment by andersmelander** [comment]
You can replace the other `MOVZX` with `MOV` too.
`MOVZX` is only used when moving a narrow type into a wider type. E.g. a BYTE into a WORD or a DWORD. The `ZX` part means _zero extend_ = fill the rest with zero. Apparently, the Delphi compiler just uses `MOV` instead of `MOVZX` if both source and target are the same sizes.
One could also use `MOV <reg>, DWORD PTR [<adr>]` to be verbose but since all the E?X registers are DWORDs it's really superfluous.

> `MOV     EDI, [EIP + GCrcTable]`

That really looks like a bug. The EIP register is the instruction pointer (which points to the currently executing asm code) and there's no way the CrcTable is located relative to that. My guess is that the statement simply should read:

`MOV     EDI, [GCrcTable]`

In other words; The same as for Delphi.

**Comment by regs01** [comment]
> You can replace the other `MOVZX` with `MOV` too.

Replace them makes it compiles without error



> MOV EDI, [GCrcTable]

x64 version has similar treatment, but compiles without errors.
```
{$IFNDEF FPC}
        MOV     RDI, [GCrcTable]
{$ELSE}
        MOV     RDI, [RIP + GCrcTable]
{$ENDIF} 
```

**Comment by andersmelander** [comment]
> x64 version has similar treatment, but compiles without errors.
> ```
> {$IFNDEF FPC}
>         MOV     RDI, [GCrcTable]
> {$ELSE}
>         MOV     RDI, [RIP + GCrcTable]
> {$ENDIF} ```

That looks correct. See [Alignment issue in x64 code, Free Pascal](https://stackoverflow.com/questions/28366918/alignment-issue-in-x64-code-free-pascal)

Unless there are any more problems with the FPC asm code, I will commit the changes and close the issue.

**Comment by andersmelander** [comment]
> You can replace the other `MOVZX` with `MOV` too.

I was wrong about these `MOVZX`:

```
  MOVZX   EBX, [EAX + 4 * ECX]
  MOVZX   EBX, [EDI + EBX]
  MOV     [EDX + 4 * ECX + $02], BL

  MOVZX   EBX, [EAX + 4 * ECX + $01]
  MOVZX   EBX, [EDI + EBX]
  MOV     [EDX + 4 * ECX + $01], BL

  MOVZX   EBX,[EAX + 4 * ECX + $02]
  MOVZX   EBX,[EDI + EBX]
  MOV     [EDX + 4 * ECX], BL

  MOVZX   EBX, [EAX + 4 * ECX + $03]
  MOV     [EDX + 4 * ECX + $03], BL
```

Looking at the 64-bit code it's "obvious" that they are moving BYTE values into DWORD registers, so `MOVZX` is correct. We just need to tell the compiler that the source is a BYTE value.
The correct fix is this:

```
  MOVZX   EBX, [EAX + 4 * ECX].BYTE
  MOVZX   EBX, [EDI + EBX].BYTE
  MOV     [EDX + 4 * ECX + $02], BL

  MOVZX   EBX, [EAX + 4 * ECX + $01].BYTE
  MOVZX   EBX, [EDI + EBX].BYTE
  MOV     [EDX + 4 * ECX + $01], BL

  MOVZX   EBX, [EAX + 4 * ECX + $02].BYTE
  MOVZX   EBX, [EDI + EBX].BYTE
  MOV     [EDX + 4 * ECX], BL

  MOVZX   EBX, [EAX + 4 * ECX + $03].BYTE
  MOV     [EDX + 4 * ECX + $03], BL
```

**Comment by andersmelander** [comment]
...and the same for the first `MOV`/`MOVZX`

**Comment by regs01** [comment]
There are couple of more LEA lines
GR32_PortableNetworkGraphic.pas:4126
GR32_PortableNetworkGraphic.pas:4127
GR32_Png.pas:1472
GR32_Png.pas:1473
